### PR TITLE
feat : Implement Hydra for Modular Configuration Management

### LINKED ISSUE #21: Implementation of Hydra for Configuration Management
Integrate Hydra to enhance configuration management by enabling modular, hierarchical, and dynamic configuration handling. This will improve flexibility, maintainability, and experiment reproducibility.

Tasks:
- Install Hydra and add it to project dependencies.
- Refactor configuration management with structured YAML files.
- Modify scripts to load configurations dynamically using Hydra.
- Test & validate configuration parsing and overrides.
- Update documentation with usage instructions.

Acceptance Criteria:
- Hydra is integrated and functional.
- Configurations are modular and override-friendly.
- Documentation is updated.

### PR DESCRIPTION
**Pull request by Rakesh-Seenu**
# For authors

image
image

## Description
Before:

- Configuration paths were hardcoded in the codebase, making it inflexible and difficult to manage.
- Each section (data, leaderboard, algorithms) had settings directly embedded in the scripts, reducing reusability.
- Modifying configurations required manual code changes, increasing maintenance effort.

Changes Implemented now:

- Integrated Hydra for dynamic configuration management.
- Created a structured config/ directory with dedicated subfolders (data, leaderboard, algorithms).
- Each section now has default.yaml files, allowing separate and modular configuration management.
- Updated the code to load configurations dynamically using Hydra instead of hardcoded paths.
- Improved flexibility, making it easier to update and experiment with configurations using overrides.

Future Enhancements:

- Implement a Data Loader to streamline data handling and improve pipeline efficiency.
- Extend configuration management for additional custom overrides and hierarchical settings.
- Improve documentation to guide usage and further optimizations..

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [ ] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [x] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (e.g. MkDocs)
- [ ] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

**Comment by Rakesh-Seenu** [comment]
@gurdeep330 Please Let me know if any changes required.

File: tests/app/data/test_markdown.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    assert os.path.exists(empty_folder)
+    assert len(os.listdir(empty_folder)) == 0
+    with pytest.raises(ValueError, match="No SBML files found in vpeleaderboard/data/empty_folder"):
+        BasicoModel(sbml_folder_path=empty_folder)
+
+def test_get_model_metadata():
+    """
+    Test the get_model_metadata method of the BasicoModel class.
+    """
+    model = BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
```
Comment (gurdeep330) [inline tests/app/data/test_markdown.py:41]: I suggest replacing `sbml_folder_path` with `sbml_file_path`. In other words, you specify the path to the model when the object is initialized.

This is also the place you'd want to use the fixture.

File: tests/app/data/test_markdown.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+"""
+A test markdown class for pytest unit testing.
+"""
+
+import os
+import pytest
+from vpeleaderboard.data.src.basico_model import BasicoModel
+from app.data.markdown_report import generate_markdown_report
+
+@pytest.fixture(name="model")
```
Comment (gurdeep330) [inline tests/app/data/test_markdown.py:10]: But I don't see it being used in the code.

File: tests/vpeleaderboard/data/test_basico_model.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    os.makedirs(empty_folder, exist_ok=True)
+
+    with pytest.raises(ValueError, match="No SBML files found in vpeleaderboard/data/empty_folder"):
+        BasicoModel(sbml_folder_path=empty_folder)
+
+def test_get_model_metadata():
+    """
+    Test the get_model_metadata method of the BasicoModel class.
+    """
+    model = BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
```
Comment (gurdeep330) [inline tests/vpeleaderboard/data/test_basico_model.py:39]: Same comment as above

File: .gitignore [hunk 4]
```diff
@@ -9,6 +9,7 @@ __pycache__/
 # Distribution / packaging
 .Python
 build/
+outputs/
```
Comment (gurdeep330) [inline .gitignore:12]: Is this needed?

File: app/data/utils/markdown.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    Args:
+        markdown_content (str): The markdown content to save.
+        output_path (str): The file path where markdown should be saved.
+    """
+    output_dir = os.path.dirname(output_path)
+    os.makedirs(output_dir, exist_ok=True)
+
+    with open(output_path, 'w', encoding='utf-8') as file:
+        file.write(markdown_content)
+    print(f"Markdown file saved to {output_path}")
```
Comment (gurdeep330) [inline app/data/utils/markdown.py:48]: logger?

File: app/leaderboard/table.py
Comment (gurdeep330) [inline app/leaderboard/table.py:1]: Looks like table.py in `app/algorithms` and `app/leaderboard` is very similar to markdown_report.py in `app/data/utils`. So why don't we merge it?

File: tests/vpeleaderboard/data/test_basico_model.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
+"""
+A test BasicoModel class for pytest unit testing.
+"""
+
+import os
+import pytest
+from vpeleaderboard.data.src.basico_model import BasicoModel
+
+@pytest.fixture(name="model")
```
Comment (gurdeep330) [inline tests/vpeleaderboard/data/test_basico_model.py:9]: Same comment as above

**Review by gurdeep330 — CHANGES_REQUESTED** [review]
Nice first attempt :+1: 

I have made a few comments. Please let me know if you need more information or clarifications.

Besides, I believe we need an updated actions workflow that automatically runs the required script and updates the table whenever a PR is merged or a commit is made to the main branch—similar to what we did here: https://github.com/VirtualPatientEngine/literatureSurvey/blob/main/.github/workflows/mkdocs-deploy.yml

File: tests/app/data/test_markdown.py (hunk 1 above)
Comment (Rakesh-Seenu) [reply tests/app/data/test_markdown.py:41]: I have renamed the sbml_folder_path with sbml_file_path.

File: tests/vpeleaderboard/data/test_basico_model.py (hunk 3 above)
Comment (Rakesh-Seenu) [reply tests/vpeleaderboard/data/test_basico_model.py:39]: I have renamed the sbml_folder_path with sbml_file_path.

File: .gitignore (hunk 4 above)
Comment (Rakesh-Seenu) [reply .gitignore:12]: no, we don't need this, I have removed it.

File: tests/vpeleaderboard/data/test_basico_model.py (hunk 6 above)
Comment (Rakesh-Seenu) [reply tests/vpeleaderboard/data/test_basico_model.py:9]: I have removed unused fixture.

File: app/data/utils/markdown.py (hunk 5 above)
Comment (Rakesh-Seenu) [reply app/data/utils/markdown.py:48]: Instead of print I have added logger

File: app/leaderboard/table.py
Comment (gurdeep330) [reply app/leaderboard/table.py:1]: @Rakesh-Seenu I am sorry. I meant _... is very similar to **markdown.py** in `app/data/utils`...._

**Review by dmccloskey — CHANGES_REQUESTED** [review]
The progress in implementing Hydra is coming along I can see 👍. I have a few comments regarding what appears to be redundancy of code. Please let me know if I am misunderstanding anything or if you have any questions on my comments.

File: tests/vpeleaderboard/data/utils/utils.py
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply tests/vpeleaderboard/data/utils/utils.py:1]: yes Douglas, it is possible to consolidate into a single util. I am working on it.

File: app/data/utils/markdown.py (hunk 5 above)
Comment (Rakesh-Seenu) [reply app/data/utils/markdown.py:48]: I have updated the code, need to commit it

File: app/data/utils/markdown.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+file with functions to generate markdown.
+'''
+
+import os
+from typing import List, Dict, Union
+import pandas as pd
+from jinja2 import Environment, FileSystemLoader
+
+
+def create_markdown(data: List[Dict[str, Union[str, int]]],
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply app/data/utils/markdown.py:11]: The create_markdown function had slight differences in how it handled input data, whereas save_markdown remained the same. To improve maintainability and consistency, I have now consolidated both functions into a single utils file. This updated implementation can handle different input types efficiently while keeping the codebase cleaner and more modular.

File: tests/app/data/test_markdown_utils.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    template_file = tmp_path / "template.md"
+    template_file.write_text(SAMPLE_TEMPLATE)
+    return str(tmp_path), "template.md"
+
+
+def test_create_markdown(temp_template):
+    """Test if markdown content is correctly generated from the template."""
+    template_dir, template_file = temp_template
+    markdown_content = create_markdown(SAMPLE_DATA, template_dir, template_file)
+    assert "Model Report" in markdown_content
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply tests/app/data/test_markdown_utils.py:39]: sure, I will add test case for current_time as well.

File: tests/app/data/test_markdown.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+def test_generate_markdown_report():
+    """
+    Test the generate_markdown_report function.
+    """
+    folder_path = "vpeleaderboard/data/models"
+    output_path = "tests/test_index.md"
+    template_dir = "app/templates"
+    template_file = "data.html"
+    assert os.path.exists(folder_path)
+    generate_markdown_report(folder_path,template_dir,template_file, output_path)
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply tests/app/data/test_markdown.py:57]: I have updated as suggested change.

**Comment by Rakesh-Seenu** [comment]
image
I have consolidated the save_markdown and create_markdown functions into a newly created utils folder inside the app directory. This restructuring ensures that all sections can efficiently utilize these functions from a centralized location, improving maintainability and consistency across the project.

File: app/leaderboard/table.py
Comment (dmccloskey) [reply app/leaderboard/table.py:1]: Wow. That really reduced the lines of code 👍.

File: tests/app/data/test_markdown.py [hunk 10]
```diff
@@ -7,20 +7,20 @@
 from vpeleaderboard.data.src.basico_model import BasicoModel
 from app.data.markdown_report import generate_markdown_report
 
-@pytest.fixture(name="model")
-def model_fixture():
-    """
-    A fixture for the BasicoModel class.
-    """
-    return BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
+# @pytest.fixture(name="model")
```
Comment (dmccloskey) [inline tests/app/data/test_markdown.py:10]: Is this comment needed?

File: tests/app/leaderboard/test_table_creation1.py
Comment (dmccloskey) [inline tests/app/leaderboard/test_table_creation1.py:1]: Appears to be a duplicate?

File: tests/vpeleaderboard/data/test_basico_model.py [hunk 11]
```diff
@@ -6,20 +6,20 @@
 import pytest
 from vpeleaderboard.data.src.basico_model import BasicoModel
 
-@pytest.fixture(name="model")
-def model_fixture():
-    """
-    A fixture for the BasicoModel class.
-    """
-    return BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
+# @pytest.fixture(name="model")
```
Comment (dmccloskey) [inline tests/vpeleaderboard/data/test_basico_model.py:9]: Is this comment needed?

File: vpeleaderboard/data/src/test/test.ipynb
Comment (dmccloskey) [inline vpeleaderboard/data/src/test/test.ipynb:1]: Is the inclusion of this file intentional?

File: app/data/markdown_report.py [hunk 12]
```diff
@@ -11,37 +11,38 @@
...
     Args:
-        folder_path (str): Path to the folder containing SBML XML files.
+        input_file (str): Path to the folder containing SBML XML files.
         template_dir (str): Path to the template directory.
         template_file (str): Template file name.
-        output_path (str): Path where the markdown report will be saved.
+        output_file (str): Path where the markdown report will be saved.
     """
-    biomodel = BasicoModel(sbml_folder_path=folder_path)
+    # input_file = str(cfg.paths.input_file.get("folder", ""))  # Extracts folder value safely
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply app/data/markdown_report.py:34]: no, we don't need it.

File: tests/app/data/test_markdown.py (hunk 10 above)
Comment (Rakesh-Seenu) [reply tests/app/data/test_markdown.py:10]: no, we don't need it. I have removed it.

File: tests/vpeleaderboard/data/test_basico_model.py (hunk 11 above)
Comment (Rakesh-Seenu) [reply tests/vpeleaderboard/data/test_basico_model.py:9]: no, we don't need it. I have removed it.

File: vpeleaderboard/data/src/test/test.ipynb
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/src/test/test.ipynb:1]: I was working with the data loader to test this file. I will remove it in the next PR.

File: tests/app/leaderboard/test_table_creation1.py
Comment (Rakesh-Seenu) [reply tests/app/leaderboard/test_table_creation1.py:1]: No, It is not duplicate. This file is used for leaderboard table generation. It is similar but differs in the Hydra configuration

File: app/utils/markdown_save.py [hunk 13]
```diff
@@ -1,20 +1,40 @@
...
+        markdown_content (str): The markdown content to save.
+        output_path (str): The file path where markdown should be saved.
+    """
+    output_dir = os.path.dirname(output_path)
+    os.makedirs(output_dir, exist_ok=True)
+    with open(output_path, 'w', encoding='utf-8') as file:
+        file.write(markdown_content)
+    logger.info("Markdown report saved at: %s", output_path)
+
+def create_markdown(data: Union[List[Dict[str, Union[str, int]]], pd.DataFrame],
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply app/utils/markdown_save.py:30]: Thanks Douglas

File: app/algorithms/table.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python3
+'''
+This script demonstrates draft of creating tables in Markdown.
+'''
+
+import os
+import pandas as pd
+import hydra
+from jinja2 import Environment, FileSystemLoader
+from app.utils import save_markdown, create_markdown
```
Comment (gurdeep330) [inline app/algorithms/table.py:10]: I recommend using relative imports; otherwise, the script will only run when executed from the root directory.

File: app/leaderboard/table.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python3
+'''
+This script demonstrates draft of creating tables in Markdown.
+'''
+
+import os
+import pandas as pd
+import hydra
+from app.utils.markdown_save import save_markdown, create_markdown
```
Comment (gurdeep330) [inline app/leaderboard/table.py:9]: I recommend using relative imports; otherwise, the script will only run when executed from the root directory.

File: app/data/markdown_report.py [hunk 16]
```diff
@@ -5,40 +5,53 @@
...
 """
 
+import sys
 import os
 import logging
 import pandas as pd
+import hydra
 from vpeleaderboard.data.src.basico_model import BasicoModel
-from vpeleaderboard.data.utils.markdown import create_markdown, save_markdown
+from app.utils import create_markdown, save_markdown
```
Comment (gurdeep330) [inline app/data/markdown_report.py:14]: I recommend using relative imports; otherwise, the script will only run when executed from the root directory.

File: tests/app/data/test_markdown.py [hunk 17]
```diff
@@ -7,20 +7,13 @@
...
-    return BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
-
 def test_validate_sbml_folder_path_success():
     """
     Test SBML directory validation when XML files exist.
     """
     assert os.path.exists("vpeleaderboard/data/models")
     assert len([f for f in os.listdir("vpeleaderboard/data/models") if f.endswith(".xml")]) > 0
-    model = BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
+    model = BasicoModel(sbml_file_path="vpeleaderboard/data/models")
```
Comment (gurdeep330) [inline tests/app/data/test_markdown.py:16]: As I mentioned in my previous review, I still think it's better to make it model-specific rather than folder-specific—one object per model.

File: tests/app/data/test_markdown.py [hunk 18]
```diff
@@ -32,13 +25,13 @@ def test_validate_sbml_folder_path_failure():
...
     with pytest.raises(ValueError, match="No SBML files found in vpeleaderboard/data/empty_folder"):
-        BasicoModel(sbml_folder_path=empty_folder)
+        BasicoModel(sbml_file_path=empty_folder)
 
 def test_get_model_metadata():
     """
     Test the get_model_metadata method of the BasicoModel class.
     """
-    model = BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
+    model = BasicoModel(sbml_file_path="vpeleaderboard/data/models")
```
Comment (gurdeep330) [inline tests/app/data/test_markdown.py:34]: Same comment as before

File: tests/vpeleaderboard/data/test_basico_model.py [hunk 19]
```diff
@@ -30,13 +23,13 @@ def test_validate_sbml_folder_path_failure():
...
     with pytest.raises(ValueError, match="No SBML files found in vpeleaderboard/data/empty_folder"):
-        BasicoModel(sbml_folder_path=empty_folder)
+        BasicoModel(sbml_file_path=empty_folder)
 
 def test_get_model_metadata():
     """
     Test the get_model_metadata method of the BasicoModel class.
     """
-    model = BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
+    model = BasicoModel(sbml_file_path="vpeleaderboard/data/models")
```
Comment (gurdeep330) [inline tests/vpeleaderboard/data/test_basico_model.py:32]: Same comment as before

**Review by gurdeep330 — CHANGES_REQUESTED** [review]
Thanks for addressing the reviews, @Rakesh-Seenu 🙏 
I still have 2 major concerns:
1. Use relative imports instead of absolute paths.  
2. The `BascoModel` class should have a single instance per model.

**Comment by dmccloskey** [comment]
> Thanks for addressing the reviews, @Rakesh-Seenu 🙏 I still have 2 major concerns:
> 
> 1. Use relative imports instead of absolute paths.
> 2. The `BascoModel` class should have a single instance per model.

@gurdeep330 my bad. I did not catch that before merging. @Rakesh-Seenu can you address these issues in a new PR? I think this will be a bit cleaner than reverting since at the moment it is only you working on the code.

**Comment by gurdeep330** [comment]
@dmccloskey, no worries. @Rakesh-Seenu and I already discussed this. Rakesh will take it in a new PR. Thx!
